### PR TITLE
[extractors/podbayfm] Add extractor

### PR DESCRIPTION
**Pull request by schnusch**
**IMPORTANT**: PRs without the template will be CLOSED

### Description of your *pull request* and other information

</details>

ADD DESCRIPTION HERE

Fixes #

<details open><summary>Template</summary> 

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8) and [ran relevant tests](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Fix or improvement to an extractor (Make sure to add/update tests)
- [x] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [ ] Core bug fix/improvement
- [ ] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

### CODE REVIEW

File: yt_dlp/extractor/podbayfm.py [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+
+    def _parse_jwt_add_padding(self, token, video_id):
+        b64 = token.split('.')[1]
+        b64 += (4 - (len(b64) % 4)) * '='
+        json = urlsafe_b64decode(b64).decode('utf-8')
+        return self._parse_json(json, video_id)
+
+    def _result_from_props(self, props, episode_id):
+        from pprint import pprint
+        pprint(props)
```
Comment (pukkandan) [inline yt_dlp/extractor/podbayfm.py:29]: ?

File: yt_dlp/extractor/podbayfm.py [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+        'params': {
+            'skip_download': True,
+        },
+    }]
+
+    def _parse_jwt_add_padding(self, token, video_id):
+        b64 = token.split('.')[1]
+        b64 += (4 - (len(b64) % 4)) * '='
+        json = urlsafe_b64decode(b64).decode('utf-8')
+        return self._parse_json(json, video_id)
```
Comment (pukkandan) [inline yt_dlp/extractor/podbayfm.py:25]: utils.jwt_decode_hs256

File: yt_dlp/extractor/podbayfm.py [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+            'timestamp': props['timestamp'],
+            'duration': int(props['duration'], 10),
+        }
+
+    def _real_extract(self, url):
+        episode_id = self._match_id(url)
+        webpage = self._download_webpage(url, episode_id)
+        data = self._parse_json(
+            self._html_search_regex(r'<script .*?\bid="__NEXT_DATA__".*?>(.*)</script>', webpage, 'JSON data'),
+            episode_id)
```
Comment (pukkandan) [inline yt_dlp/extractor/podbayfm.py:46]: _search_nextjs_data

File: yt_dlp/extractor/podbayfm.py [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+        props = data['props']['pageProps']['episode']
+        return self._result_from_props(props, episode_id)
```
Comment (pukkandan) [inline yt_dlp/extractor/podbayfm.py:48]:
```suggestion
        return self._result_from_props(data['props']['pageProps']['episode'], episode_id)
```

File: yt_dlp/extractor/podbayfm.py [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+        pprint(props)
+        image_props = self._parse_jwt_add_padding(props['image'], episode_id)
+        return {
+            'id': props.get('podcast_id') or episode_id,
+            'title': props['title'],
+            'url': props['mediaURL'],
+            'ext': 'mp3',
+            'thumbnail': image_props['url'],
+            'timestamp': props['timestamp'],
+            'duration': int(props['duration'], 10),
```
Comment (pukkandan) [inline yt_dlp/extractor/podbayfm.py:38]: make non-fatal. Read contributing guidelines

File: yt_dlp/extractor/podbayfm.py [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+            episode_id)
+        props = data['props']['pageProps']['episode']
+        return self._result_from_props(props, episode_id)
+
+
+class PodbayFMChannelIE(PodbayFMIE):
+    _VALID_URL = r'https?://podbay\.fm/p/(?P<id>[^/]*)/?(?:[\?#].*)?$'
+    _TESTS = [{
+        'url': 'https://podbay.fm/p/behind-the-bastards',
+    }]
```
Comment (pukkandan) [inline yt_dlp/extractor/podbayfm.py:55]: ??

File: yt_dlp/extractor/podbayfm.py [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+        'url': 'https://podbay.fm/p/behind-the-bastards/e/1647338400',
+        'md5': 'be0fd3275a12e9acf6d89037b9dbea04',
+        'info_dict': {
+            'id': '1647338400',
+            'ext': 'mp3',
+            'title': 'Part One: Kissinger',
+            'duration': 5001,
+        },
+        'params': {
+            'skip_download': True,
```
Comment (pukkandan) [inline yt_dlp/extractor/podbayfm.py:17]: why?

File: yt_dlp/extractor/podbayfm.py [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+                'https://podbay.fm/api/podcast?reverse=true&page=' + str(pagenum) + '&slug=' + channel_id,
+                channel_id)['podcast']
+            if not page['episodes']:
+                break
+
+            if title is None:
+                title = page['title']
+
+            for episode in page['episodes']:
+                episodes.append(self._result_from_props(episode, None))
```
Comment (pukkandan) [inline yt_dlp/extractor/podbayfm.py:75]: convert to a PagedList

**Review by pukkandan — CHANGES_REQUESTED** [review]

File: yt_dlp/extractor/podbayfm.py [hunk 9]
```diff
@@ -1,82 +1,76 @@
...
+import functools
+
 from .common import InfoExtractor
+from ..utils import OnDemandPagedList, jwt_decode_hs256, try_get
+
+
+def fix_jwt_padding(token):
+    token = token.split('.')
+    token[1] += (4 - (len(token[1]) % 4)) * '='
+    return '.'.join(token)
```
Comment (pukkandan) [inline yt_dlp/extractor/podbayfm.py:10]: You can instead improve `jwt_decode_hs256` with this code

File: yt_dlp/extractor/podbayfm.py [hunk 10]
```diff
@@ -1,82 +1,76 @@
...
+
+
+def result_from_props(props, episode_id):
+    image_props = jwt_decode_hs256(fix_jwt_padding(props['image']))
+    return {
+        'id': props.get('podcast_id') or episode_id,
+        'title': props['title'],
```
Comment (pukkandan) [inline yt_dlp/extractor/podbayfm.py:17]: .get

File: yt_dlp/extractor/podbayfm.py [hunk 11]
```diff
@@ -1,82 +1,76 @@
...
+        'url': props['mediaURL'],
+        'ext': 'mp3',
+        'thumbnail': image_props.get('url'),
+        'timestamp': props.get('timestamp'),
+        'duration': try_get(props, lambda p: int(p['duration'], 10))
```
Comment (pukkandan) [inline yt_dlp/extractor/podbayfm.py:22]: int_or_none

File: yt_dlp/extractor/podbayfm.py [hunk 12]
```diff
@@ -1,82 +1,76 @@
...
+            'title': 'Behind the Bastards',
+        },
     }]
+    _PAGE_SIZE = 10
+
+    def _fetch_page(self, channel_id, pagenum):
+        page = self._download_json(
+            'https://podbay.fm/api/podcast?reverse=true&page=' + str(pagenum) + '&slug=' + channel_id,
+            channel_id)['podcast']
+        return [result_from_props(e, None) for e in page['episodes']]
```
Comment (pukkandan) [inline yt_dlp/extractor/podbayfm.py:64]: Normally, you'd return a url_result. Here you are returing the direct video result, and I can see that makes sense. But when you do this, you need to explicitly set the correct `webpage_url`, `extractor_key` and `extractor`

File: yt_dlp/extractor/podbayfm.py [hunk 13]
```diff
@@ -1,82 +1,76 @@
...
+        'info_dict': {
+            'id': 'behind-the-bastards',
+            'title': 'Behind the Bastards',
+        },
     }]
+    _PAGE_SIZE = 10
+
+    def _fetch_page(self, channel_id, pagenum):
+        page = self._download_json(
+            'https://podbay.fm/api/podcast?reverse=true&page=' + str(pagenum) + '&slug=' + channel_id,
```
Comment (pukkandan) [inline yt_dlp/extractor/podbayfm.py:62]: Use fstring

File: yt_dlp/extractor/podbayfm.py [hunk 14]
```diff
@@ -1,82 +1,76 @@
...
-                break
-
-            if title is None:
-                title = page['title']
-
-            for episode in page['episodes']:
-                episodes.append(self._result_from_props(episode, None))
+        first_page = self._download_json(
+            'https://podbay.fm/api/podcast?reverse=true&page=0&slug=' + channel_id,
+            channel_id)['podcast']
```
Comment (pukkandan) [inline yt_dlp/extractor/podbayfm.py:74]: pass this page into `entries` so as to not re-download it

File: yt_dlp/extractor/podbayfm.py [hunk 15]
```diff
@@ -1,82 +1,76 @@
...
 
-        return {
-            '_type': 'playlist',
-            'id': channel_id,
-            'title': title,
-            'entries': episodes,
-        }
+        return self.playlist_result(entries, channel_id, first_page['title'])
```
Comment (pukkandan) [inline yt_dlp/extractor/podbayfm.py:76]: .get

**Review by pukkandan — CHANGES_REQUESTED** [review]

**Comment by pukkandan** [comment]
Pls avoid force-pushing

File: yt_dlp/extractor/podbayfm.py [hunk 16]
```diff
@@ -0,0 +1,80 @@
+from .common import InfoExtractor
+from ..utils import OnDemandPagedList, int_or_none, jwt_decode_hs256
+
+
+def result_from_props(props, episode_id):
```
Comment (pukkandan) [inline yt_dlp/extractor/podbayfm.py:5]:
```suggestion
def result_from_props(props, episode_id=None):
```

File: yt_dlp/extractor/podbayfm.py [hunk 17]
```diff
@@ -0,0 +1,80 @@
...
+    _PAGE_SIZE = 10
+
+    def _fetch_page(self, channel_id, pagenum):
+        return self._download_json(
+            f'https://podbay.fm/api/podcast?reverse=true&page={pagenum}&slug={channel_id}',
+            channel_id)['podcast']
+
+    @staticmethod
+    def _results_from_page(channel_id, page):
+        return [result_from_props(e, None) | {
```
Comment (pukkandan) [inline yt_dlp/extractor/podbayfm.py:60]:
```suggestion
        return [{
            **result_from_props(e),
```
`|` is unsupported in 3.7

File: yt_dlp/extractor/podbayfm.py [hunk 18]
```diff
@@ -0,0 +1,80 @@
...
+            'extractor': 'PodbayFM',
+            'extractor_key': 'PodbayFM',
```
Comment (pukkandan) [inline yt_dlp/extractor/podbayfm.py:62]:
```suggestion
            'extractor': PodbayFMIE.IE_NAME,
            'extractor_key': PodbayFM.ie_key(),
```

File: yt_dlp/extractor/podbayfm.py [hunk 19]
```diff
@@ -0,0 +1,80 @@
...
+
+        entries = OnDemandPagedList(
+            lambda pagenum:
+                self._results_from_page(channel_id, self._fetch_page(channel_id, pagenum)),
+            self._PAGE_SIZE)
+
+        first_page = self._fetch_page(channel_id, 0)
+        # add it to entries's cache
+        if entries._use_cache:
+            entries._cache[0] = self._results_from_page(channel_id, first_page)
```
Comment (pukkandan) [inline yt_dlp/extractor/podbayfm.py:78]:
```suggestion
        first_page = self._fetch_page(channel_id, 0)
        entries = OnDemandPagedList(
            lambda pagenum: self._results_from_page(
                channel_id, self._fetch_page(channel_id, pagenum) if pagenum else first_page),
            self._PAGE_SIZE)
```

File: yt_dlp/utils.py [hunk 20]
```diff
@@ -5424,6 +5424,7 @@ def jwt_encode_hs256(payload_data, key, headers={}):
 # can be extended in future to verify the signature and parse header and return the algorithm used if it's not HS256
 def jwt_decode_hs256(jwt):
     header_b64, payload_b64, signature_b64 = jwt.split('.')
+    payload_b64 += (4 - (len(payload_b64) % 4)) * '='
     payload_data = json.loads(base64.urlsafe_b64decode(payload_b64))
```
Comment (pukkandan) [inline yt_dlp/utils.py:5428]:
```suggestion
    payload_data = json.loads(base64.urlsafe_b64decode(f'{payload_b64}==='))
```
should work afair

**Review by pukkandan — CHANGES_REQUESTED** [review]

File: yt_dlp/utils.py (hunk 20 above)
Comment (schnusch) [reply yt_dlp/utils.py:5428]: I does not work for me.
```$ ./yt-dlp.sh -Jv 'https://podbay.fm/p/behind-the-bastards'
[debug] Command-line config: ['-Jv', 'https://podbay.fm/p/behind-the-bastards']
[debug] Encodings: locale UTF-8, fs utf-8, pref UTF-8, out utf-8, error utf-8, screen utf-8
[debug] yt-dlp version 2022.09.01 [5d7c7d656] (source)
[debug] Lazy loading extractors is disabled
[debug] Plugins: ['SamplePluginIE', 'SamplePluginPP']
[debug] Git HEAD: 4d8f5e2cb
[debug] Python 3.9.13 (CPython 64bit) - Linux-5.15.62-x86_64-with-glibc2.34 (glibc 2.34)
[debug] Checking exe version: ffmpeg -bsfs
[debug] Checking exe version: ffprobe -bsfs
[debug] exe versions: ffmpeg 5.0.1 (setts), ffprobe 5.0.1
[debug] Optional libraries: sqlite3-2.6.0
[debug] Proxy map: {}
[debug] Loaded 1674 extractors
[debug] [PodbayFMChannel] Extracting URL: https://podbay.fm/p/behind-the-bastards
[PodbayFMChannel] behind-the-bastards: Downloading JSON metadata
[download] Downloading playlist: Behind the Bastards
ERROR: Incorrect padding
Traceback (most recent call last):
  File "/home/schnusch/Desktop/yt-dlp/yt_dlp/YoutubeDL.py", line 1477, in wrapper
    return func(self, *args, **kwargs)
  File "/home/schnusch/Desktop/yt-dlp/yt_dlp/utils.py", line 2992, in <lambda>
    return type(self.ydl)._handle_extraction_exceptions(lambda _, i: self._entries[i])(self.ydl, i)
  File "/home/schnusch/Desktop/yt-dlp/yt_dlp/utils.py", line 2827, in __getitem__
    entries = self.getslice(idx, idx + 1)
  File "/home/schnusch/Desktop/yt-dlp/yt_dlp/utils.py", line 2818, in getslice
    return list(self._getslice(start, end))
  File "/home/schnusch/Desktop/yt-dlp/yt_dlp/utils.py", line 2853, in _getslice
    page_results = self.getpage(pagenum)
  File "/home/schnusch/Desktop/yt-dlp/yt_dlp/utils.py", line 2812, in getpage
    page_results = [] if pagenum > self._pagecount else list(self._pagefunc(pagenum))
  File "/home/schnusch/Desktop/yt-dlp/yt_dlp/extractor/podbayfm.py", line 74, in <lambda>
    self._results_from_page(
  File "/home/schnusch/Desktop/yt-dlp/yt_dlp/extractor/podbayfm.py", line 60, in _results_from_page
    return [{
  File "/home/schnusch/Desktop/yt-dlp/yt_dlp/extractor/podbayfm.py", line 61, in <listcomp>
    **result_from_props(e),
  File "/home/schnusch/Desktop/yt-dlp/yt_dlp/extractor/podbayfm.py", line 7, in result_from_props
    thumbnail = jwt_decode_hs256(image_jwt).get('url') if image_jwt else None
  File "/home/schnusch/Desktop/yt-dlp/yt_dlp/utils.py", line 5428, in jwt_decode_hs256
    payload_data = json.loads(f'{base64.urlsafe_b64decode(payload_b64)}===')
  File "/nix/store/pn7863n7s2p66b0gazcylm6cccdwpzaf-python3-3.9.13/lib/python3.9/base64.py", line 133, in urlsafe_b64decode
    return b64decode(s)
  File "/nix/store/pn7863n7s2p66b0gazcylm6cccdwpzaf-python3-3.9.13/lib/python3.9/base64.py", line 87, in b64decode
    return binascii.a2b_base64(s)
binascii.Error: Incorrect padding```

Comment (bashonly) [reply yt_dlp/utils.py:5428]: the line in your log:
```py
payload_data = json.loads(f'{base64.urlsafe_b64decode(payload_b64)}===')
```
is not the same as the line that was suggested:
```py
payload_data = json.loads(base64.urlsafe_b64decode(f'{payload_b64}==='))
```

Comment (schnusch) [reply yt_dlp/utils.py]: You're right.

File: yt_dlp/extractor/podbayfm.py (hunk 19 above)
Comment (schnusch) [reply yt_dlp/extractor/podbayfm.py]: I agree using private properties is ugly, but if it's common there should maybe be a proper way to add to `PagedList._cache`.

Comment (pukkandan) [reply yt_dlp/extractor/podbayfm.py]: what is wrong with the current method?

File: yt_dlp/extractor/podbayfm.py [hunk 21]
```diff
@@ -0,0 +1,79 @@
...
+
+def result_from_props(props, episode_id=None):
+    image_jwt = props.get('image')
+    thumbnail = jwt_decode_hs256(image_jwt).get('url') if image_jwt else None
+    return {
+        'id': props.get('podcast_id') or episode_id,
+        'title': props.get('title'),
+        'url': props['mediaURL'],
+        'ext': 'mp3',
+        'thumbnail': thumbnail,
```
Comment (pukkandan) [inline yt_dlp/extractor/podbayfm.py:13]:
```suggestion
    return {
        'id': props.get('podcast_id') or episode_id,
        'title': props.get('title'),
        'url': props['mediaURL'],
        'ext': 'mp3',
        'thumbnail': try_call(lambda: jwt_decode_hs256(props['image'])['url']),
```

File: yt_dlp/extractor/podbayfm.py [hunk 22]
```diff
@@ -0,0 +1,79 @@
...
+
+    def _real_extract(self, url):
+        channel_id = self._match_id(url)
+
+        first_page = self._fetch_page(channel_id, 0)
+        entries = OnDemandPagedList(
+            lambda pagenum:
+                self._results_from_page(
+                    channel_id,
+                    first_page if pagenum == 0 else self._fetch_page(channel_id, pagenum)),
```
Comment (pukkandan) [inline yt_dlp/extractor/podbayfm.py:76]:
```suggestion
        entries = OnDemandPagedList(
            lambda pagenum: self._results_from_page(
                channel_id, self._fetch_page(channel_id, pagenum) if pagenum else first_page),
```

**Review by pukkandan — APPROVED** [review]

File: yt_dlp/extractor/podbayfm.py (hunk 19 above)
Comment (schnusch) [reply yt_dlp/extractor/podbayfm.py]: It's alright, but `PagedList._cache` seems to be made for this.

**Comment by pukkandan** [comment]
pls fix linter and run the tests

**Comment by pukkandan** [comment]
In future, pls avoid force-pushing

**Pull request merged**